internal/charmstore/resources_test.go

```diff
@@ -175,6 +177,222 @@ func (s *resourceSuite) TestListResourcesResourceNotFound(c *gc.C) {
+	size:        fakeBlobSize,
+	expectError: `cannot put archive blob: hash mismatch`,
+}}
+
+func (s *resourceSuite) TestUploadResourceErrors(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, _ := addCharm(c, store, curl)
+	for i, test := range uploadResourceErrorTests {
+		c.Logf("%d. %s", i, test.about)
+		_, err := store.UploadResource(entity, test.name, strings.NewReader(test.blob), test.hash, test.size)
+		c.Assert(err, gc.ErrorMatches, test.expectError)
+	}
+}
+
+var resolveResourceTests = []struct {
+	about          string
+	name           string
+	revision       int
+	channel        params.Channel
+	expectResource int
+	expectError    string
+}{{
+	about:          "revision specified without channel",
+	name:           "for-install",
+	revision:       0,
+	channel:        params.NoChannel,
+	expectResource: 0,
+}, {
+	about:          "revision specified on stable channel",
+	name:           "for-install",
+	revision:       1,
+	channel:        params.StableChannel,
+	expectResource: 1,
+}, {
+	about:          "revision specified on development channel",
+	name:           "for-install",
+	revision:       2,
+	channel:        params.DevelopmentChannel,
+	expectResource: 2,
+}, {
+	about:       "revision specified that doesn't exist",
+	name:        "for-install",
+	revision:    3,
+	channel:     params.UnpublishedChannel,
+	expectError: `cs:~charmers/xenial/starsay-3 has no "for-install" revision 3 resource`,
+}, {
+	about:          "no revision specified without channel",
+	name:           "for-install",
+	revision:       -1,
+	channel:        params.NoChannel,
+	expectResource: 0,
+}, {
+	about:          "no revision specified on stable channel",
+	name:           "for-install",
+	revision:       -1,
+	channel:        params.StableChannel,
+	expectResource: 0,
+}, {
+	about:          "no revision specified on development channel",
+	name:           "for-install",
+	revision:       -1,
+	channel:        params.DevelopmentChannel,
+	expectResource: 1,
+}, {
+	about:          "no revision specified on unpublished channel",
+	name:           "for-install",
+	revision:       -1,
+	channel:        params.UnpublishedChannel,
+	expectResource: 2,
+}, {
+	about:       "no resource with name",
+	name:        "for-setup",
+	revision:    -1,
+	channel:     params.UnpublishedChannel,
+	expectError: `cs:~charmers/xenial/starsay-3 has no "for-setup" resource`,
+}}
+
+func (s *resourceSuite) TestResolveResource(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, _ := addCharm(c, store, curl)
+
+	var resources []*mongodoc.Resource
+	for _, ch := range []params.Channel{
+		params.StableChannel,
+		params.DevelopmentChannel,
+		params.NoChannel,
+	} {
+		res, err := store.UploadResource(entity, "for-install", strings.NewReader("fake content"), fakeBlobHash, fakeBlobSize)
```

Comment:
I'm wondering if we should have similar logic to the archive upload, where if you upload the same content as the latest revision, it doesn't increment the revision number (that would mean you'd need to use a different blob for each upload I guess)